src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
             {
                 return this.ParseTupleType();
             }
+            else if (IsFunctionPointerStart())
+            {
+                return ParseFunctionPointerTypeSyntax();
+            }
             else
             {
                 var name = this.CreateMissingIdentifierName();
                 return this.AddError(name, ErrorCode.ERR_TypeExpected);
             }
         }
 
+#nullable enable
+        private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
+        {
+            Debug.Assert(IsFunctionPointerStart());
+            var @delegate = EatToken(SyntaxKind.DelegateKeyword);
+            var asterisk = EatToken(SyntaxKind.AsteriskToken);
+            // Invalid calling conventions will be reported at type binding time
+            var callingConvention = IsPossibleCallingConvention() ? EatTokenAsKind(SyntaxKind.IdentifierToken, reportError: false) : null;
+
+            if (!IsPossibleFunctionPointerParameterListStart(CurrentToken))
+            {
+                var lessThanTokenError = WithAdditionalDiagnostics(SyntaxFactory.MissingToken(SyntaxKind.LessThanToken), GetExpectedTokenError(SyntaxKind.LessThanToken, SyntaxKind.None));
+                var typesError = _pool.AllocateSeparated<ModifiedType>();
+                var identifierError = AddError(CreateMissingIdentifierName(), ErrorCode.ERR_TypeExpected);
+                var typeError = SyntaxFactory.ModifiedType(default, identifierError);
+                typesError.Add(typeError);
+                var greaterThanTokenError = WithAdditionalDiagnostics(SyntaxFactory.MissingToken(SyntaxKind.GreaterThanToken), GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None));
```

Comment:
>GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None) [](start = 127, length = 67)

Also looks like an unnecessary error. What errors are we going to report for ```delegate* >```? Are we going to say ```>``` is expected at the ```>``` character?

Reply:
I have a test for this, and yes, we do report a `>` missing character on that token.

---
In reply to: [349376764](https://github.com/dotnet/roslyn/pull/39879#discussion_r349376764) [](ancestors = 349376764)

Reply:
I'll remove the extra errors.

---
In reply to: [349377335](https://github.com/dotnet/roslyn/pull/39879#discussion_r349377335) [](ancestors = 349377335,349376764)